### PR TITLE
feat: implemented hooks system

### LINKED ISSUE #11: Implementing event emitters
Hey @mateonunez , wanted to share with you something I had in my mind for quite some time.
What about bringing some event emission inside the expirable data structures? That would provide a wider usage of those.

What do you think? Any idea about the implementation?

### PR DESCRIPTION
**Pull request by Cadienvan**
As mentioned and discussed in #11 , here's a proposal for implementing hooks.

I'll tag @mateonunez as we discussed it together and would like to also have its review with @rollsappletree's one.

### CODE REVIEW

File: src/Queue/index.test.ts [hunk 1]
```diff
@@ -56,3 +56,23 @@ describe('ExpirableQueue', () => {
     expect(queue.size).toBe(0);
   });
 });
+
+describe('ExpirableQueue hooks', () => {
+  it('should call the beforeExpire hook before expiring an entry', () => {
+    const queue = new ExpirableQueue([1, 2, 3], { defaultTtl: 10 });
+    const beforeExpire = jest.fn();
+    queue.addHook('beforeExpire', beforeExpire);
```
Comment (mateonunez) [inline src/Queue/index.test.ts:64]: You can check if the item is not yet expired.

File: src/Queue/index.test.ts [hunk 2]
```diff
@@ -56,3 +56,23 @@ describe('ExpirableQueue', () => {
...
+    sleep(20);
+    jest.advanceTimersByTime(20);
+    expect(beforeExpire).toHaveBeenCalledTimes(3);
+  });
+
+  it('should call the afterExpire hook after expiring an entry', () => {
+    const queue = new ExpirableQueue([1, 2, 3], { defaultTtl: 10 });
+    const afterExpire = jest.fn();
+    queue.addHook('afterExpire', afterExpire);
```
Comment (mateonunez) [inline src/Queue/index.test.ts:74]: You can verify that the item has expired.

**Review by mateonunez — COMMENTED** [review]
The code looks good. I'd just add specific tests to check if the item has expired or not in hooks.

Furthermore, it's a good idea, in your opinion, to return the key of the item? Some methods use it to clear the timeout or to delete the element. WDYT?

File: src/Queue/index.test.ts (hunk 1 above)
Comment (Cadienvan) [reply src/Queue/index.test.ts:64]: You are right, done!

File: src/Queue/index.test.ts (hunk 2 above)
Comment (Cadienvan) [reply src/Queue/index.test.ts:74]: You are right, done!

**Comment by Cadienvan** [comment]
Both tests updated as your questions are pretty valid, also added key to the hook return, can't know how people could interact with our library!

**Review by mateonunez — APPROVED** [review]
Looks perfect! 

LGTM

**Review by rollsappletree — APPROVED** [review]

**Pull request merged**